src/contract.rs

```diff
@@ -5,132 +5,158 @@ use crate::balance::{read_state, write_state};
-        Domain::Transfer,
-        (to.clone(), amount.clone()).into_env_val(&e),
-    );
-    spend_balance(&e, from_id, amount.clone());
-    receive_balance(&e, to, amount);
-}
+    fn xfer(e: Env, from: KeyedAuthorization, to: Identifier, amount: BigInt);
 
-#[contractfn]
-pub fn xfer_from(
-    e: Env,
-    spender: KeyedAuthorization,
-    from: Identifier,
-    to: Identifier,
-    amount: BigInt,
-) {
-    let spender_id = spender.get_identifier(&e);
-    check_auth(
-        &e,
-        spender,
-        Domain::TransferFrom,
-        (from.clone(), to.clone(), amount.clone()).into_env_val(&e),
+    fn xfer_from(
+        e: Env,
+        spender: KeyedAuthorization,
+        from: Identifier,
+        to: Identifier,
+        amount: BigInt,
     );
-    spend_allowance(&e, from.clone(), spender_id, amount.clone());
-    spend_balance(&e, from, amount.clone());
-    receive_balance(&e, to, amount);
-}
 
-#[contractfn]
-pub fn burn(e: Env, admin: Authorization, from: Identifier, amount: BigInt) {
-    let auth = to_administrator_authorization(&e, admin);
-    check_auth(
-        &e,
-        auth,
-        Domain::Burn,
-        (from.clone(), amount.clone()).into_env_val(&e),
-    );
-    spend_balance(&e, from, amount);
-}
+    fn burn(e: Env, admin: Authorization, from: Identifier, amount: BigInt);
 
-#[contractfn]
-pub fn freeze(e: Env, admin: Authorization, id: Identifier) {
-    let auth = to_administrator_authorization(&e, admin);
-    check_auth(
-        &e,
-        auth,
-        Domain::Freeze,
-        (id.clone(),).clone().into_env_val(&e),
-    );
-    write_state(&e, id, true);
-}
+    fn freeze(e: Env, admin: Authorization, id: Identifier);
 
-#[contractfn]
-pub fn mint(e: Env, admin: Authorization, to: Identifier, amount: BigInt) {
-    let auth = to_administrator_authorization(&e, admin);
-    check_auth(
-        &e,
-        auth,
-        Domain::Mint,
-        (to.clone(), amount.clone()).into_env_val(&e),
-    );
-    receive_balance(&e, to, amount);
-}
+    fn mint(e: Env, admin: Authorization, to: Identifier, amount: BigInt);
 
-#[contractfn]
-pub fn set_admin(e: Env, admin: Authorization, new_admin: Identifier) {
-    let auth = to_administrator_authorization(&e, admin);
-    check_auth(
-        &e,
-        auth,
-        Domain::SetAdministrator,
-        (new_admin.clone(),).into_env_val(&e),
-    );
-    write_administrator(&e, new_admin);
+    fn set_admin(e: Env, admin: Authorization, new_admin: Identifier);
+
+    fn unfreeze(e: Env, admin: Authorization, id: Identifier);
 }
 
-#[contractfn]
-pub fn unfreeze(e: Env, admin: Authorization, id: Identifier) {
-    let auth = to_administrator_authorization(&e, admin);
-    check_auth(&e, auth, Domain::Unfreeze, (id.clone(),).into_env_val(&e));
-    write_state(&e, id, false);
+pub struct Token;
+
+#[contractimpl]
+impl TokenTrait for Token {
+    fn initialize(e: Env, admin: Identifier) {
+        if has_administrator(&e) {
+            panic!()
```

Comment:
Note that you can add messages to panics. The messages will be erased as part of the long nightly build command you've got in your Makefile. When you run tests however you'll get a nice message, and it might improve readability for folks to see the error message here.
```suggestion
            panic!("contract does not have administrator defined")
```

I think it would be interesting for us to explore how we could encode a very small compact code into the panic message, stop erasing it, and somehow have that translated into a `Status` code in the future. cc @graydon @jayz22

Reply:
Opened https://github.com/stellar/rs-stellar-token-contract/issues/17